test/jdk/sun/security/pkcs12/KeytoolOpensslInteropTest.java

```diff
@@ -23,57 +23,116 @@
+ *          some openssl keystores, java operates on it and vice versa.
+ *
+ *          Note: This test executes some openssl command, so need to set
+ *          openssl root path using system property "test.openssl.path".
+ *          Install OpenSSL 1.1.* version and set this variable.
+ *
  * @modules java.base/sun.security.pkcs
  *          java.base/sun.security.util
- * @summary Customizing the generation of a PKCS12 keystore
+ * @library /test/lib
+ * @library /sun/security/pkcs11/
+ * @run main/othervm/timeout=600 KeytoolOpensslInteropTest
  */
 
 import jdk.test.lib.Asserts;
 import jdk.test.lib.SecurityTools;
 import jdk.test.lib.process.OutputAnalyzer;
 
 import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.io.UncheckedIOException;
 import java.nio.file.DirectoryStream;
-import java.nio.file.Files;
-import java.nio.file.Path;
 import java.security.KeyStore;
 import java.util.Base64;
 import java.util.Objects;
 
+import jdk.test.lib.Platform;
+import jdk.test.lib.SecurityTools;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.artifacts.Artifact;
+import jdk.test.lib.artifacts.ArtifactResolver;
+import jdk.test.lib.artifacts.ArtifactResolverException;
+
 import static jdk.test.lib.security.DerUtils.*;
 import static sun.security.util.KnownOIDs.*;
 import static sun.security.pkcs.ContentInfo.*;
 
-public class ParamsTest  {
+public class KeytoolOpensslInteropTest {
 
     public static void main(String[] args) throws Throwable {
-
-        // De-BASE64 textual files in ./params to `pwd`
-        try (DirectoryStream<Path> stream = Files.newDirectoryStream(
-                Path.of(System.getProperty("test.src"), "params"),
-                p -> !p.getFileName().toString().equals("README"))) {
-            stream.forEach(p -> {
-                try (InputStream is = Files.newInputStream(p);
-                     OutputStream os = Files.newOutputStream(p.getFileName())) {
-                    Base64.getMimeDecoder().wrap(is).transferTo(os);
-                } catch (IOException e) {
-                    throw new UncheckedIOException(e);
-                }
-            });
+        String opensslPath = System.getProperty("test.openssl.path");
+        if (opensslPath == null) {
+            opensslPath = fetchOpenssl();
+        }
+        if (opensslPath != null) {
+            opensslPath = opensslPath + File.separator + "bin"
+                + File.separator + "openssl";
+            generateInitialKeystores(opensslPath);
+            testWithJavaCommands();
+            testWithOpensslCommands(opensslPath);
+        } else {
```

Comment:
I still think it's better to succeed with a warning when the openssl binary cannot be found. IMHO it's a little unfriendly to force people setting up a system property to let the test pass. It's also dependent on runner machines and the user might have to adjust their scripts or launcher all the time. I would rather keep the old test/data if I want to ensure the test gets run everywhere.

Also, it might also help if the test can search for openssl, maybe simply from `/usr/bin/openssl` or `/usr/local/bin/openssl`, but this means you might need to check for the version.

Third, is it OK to let the system property pointing to the binary itself directly? When I was trying out this test I was using the openssl I built and it's not in a `bin` sub-directory.

Reply:
Ok, changed openssl selection flow as below:
1. Check whether test.openssl.path is set and it's the the preferred version(1.1.*) of openssl
2. If above property doesn't set, then look for already installed openssl (version 1.1.*) in system path /usr/bin/openssl or /usr/local/bin/openssl
3. if above is also not available try to download openssl from artifactory
If any of above 3 succeeds then perform all tests, otherwise skip all openssl command dependent tests.